### ISSUE TITLE
Support flexible tensor layouts in Triton and CuPy neurons under torch.compile

### ISSUE THREAD

**Issue by AllenYolk** [fangwei123456/SpikingFlow#753]
## Problem

SpikingJelly's multi-step Triton and CuPy neuron kernels currently require contiguous tensors. Their wrappers call `contiguous()` for inputs, state, saved intermediates, and gradients. The CuPy functional path additionally calls `flatten(1)` before entering its custom op. Under `torch.compile`, Inductor may choose channels-last layouts for convolutions. This can create repeated layout conversions at Conv/BN -> neuron boundaries and can also make Inductor select slower convolution kernels.

A controlled SEW-ResNet18 inference profile (T=4, batch 32, FP32, 136x136, RTX 4090) showed:

| mode | total GPU time | convolution | layout conversion | LIF |
|---|---:|---:|---:|---:|
| eager + Triton | 38.079 ms | 18.874 ms | 0 | 4.300 ms |
| default compile + Triton | 47.923 ms | 33.411 ms | 2.794 ms / 195 launches | 4.650 ms |
| compile + Triton, `layout_optimization=False` | 36.866 ms | 20.995 ms | 0 | 4.010 ms |

Disabling layout optimization changed compile/eager speedup from 0.800x to 1.040x. The immediate documented workaround is:

```python
compiled_model = torch.compile(
    model,
    backend="inductor",
    options={"layout_optimization": False},
)
```

The CuPy backend has the same layout conflict. A minimal Conv2d -> CuPy LIF experiment on an RTX 4090 with PyTorch 2.7.1 and CuPy 13.6.0 showed that default Inductor compilation produced a channels-last convolution result with stride `(8192, 1, 256, 8)`, followed by a generated `triton_poi_fused_view` kernel that converted it before the CuPy LIF kernel. Across three profiled iterations, `layout_optimization=False` removed that view kernel and the two convolution layout-conversion kernels; the unrelated final-state clone remained.

CuPy also has a separate compiler-capture limitation: the same minimal network fails with `torch.compile(..., fullgraph=True)` because `surrogate_registry._resolve_cuda_code_id` enters a Python `threading.Lock`. With `fullgraph=False`, Dynamo graph-breaks around this code and compiles several segments. Resolving the registry value outside the compiled forward path is a prerequisite for a fair whole-network CuPy comparison.

This issue tracks a proper flexible-layout implementation and is intentionally deferred from the current performance work.

## Why this should be feasible

The neuron update is elementwise across neurons and recurrent only across `T`. For a dense time-major tensor, a channels-last time slice remains physically dense, and the same physical offset identifies the same neuron at every time step. The core Triton and CuPy neuron equations therefore should not require a redesign.

The difficult part is the layout contract around the kernel:

- Preserve a compatible stride order for input, initial/final state, spikes, voltage sequences, and saved intermediates.
- Handle backward gradients whose stride order may differ.
- Make the outer Triton custom op, the inner wrapped Triton HOP, and the opaque CuPy custom op use accurate layout contracts.
- Move CuPy's logical `flatten(1)` out of the Conv -> custom-op boundary so a dense channels-last input does not have to be reordered first.
- Avoid a process-wide private Inductor option that changes unrelated user Triton kernels.

The public `torch.library.triton_op` API currently does not expose a per-op layout tag, and wrapped Triton kernels use Inductor's fixed-stride default. An upstream PyTorch change may be preferable to copying `triton_op` internals or maintaining a private Inductor lowering in SpikingJelly. CuPy uses opaque `torch.library.custom_op` wrappers and needs a separate, explicit layout contract.

## Proposed investigation

1. Build an inference-only Triton LIF prototype that accepts contiguous and channels-last dense time-major tensors without copying.
2. Verify whether a per-op PyTorch/Inductor layout contract is sufficient for a pure channels-last Conv -> LIF graph. If the public API is insufficient, prepare the smallest upstream PyTorch change.
3. Extend the proven approach to training, saved voltage sequences, mixed precision, and gradients.
4. Remove the CuPy surrogate-registry graph break, then prototype an N-dimensional CuPy LIF custom op that physically linearizes dense time slices without an external `flatten(1)` copy.
5. Apply each proven backend-specific mechanism to its IF/LIF/PLIF and other supported neuron kernels only after LIF shows an end-to-end benefit.
6. Keep NCHW available: the best convolution layout is workload- and GPU-dependent, so Inductor should remain free to select the measured winner.

## Acceptance criteria

- Numerical equivalence for contiguous and channels-last inputs in inference and training, including `store_v_seq`, state carry-over, and every dtype supported by each backend.
- No repeated NCHW/channels-last conversions at Conv/BN -> Triton or CuPy neuron boundaries in the generated graph/profile.
- CuPy LIF can participate in the intended whole-network compile path without a surrogate-registry graph break.
- No meaningful eager regression for either backend on existing contiguous workloads.
- Same-host end-to-end benchmarks for representative convolutional SNNs and batch sizes, reporting neuron, convolution, layout-conversion, and graph-break evidence separately for Triton and CuPy.
- Linear-only SNNs remain a control: they do not have the NCHW/channels-last convolution-layout conflict, although arbitrary non-contiguous tensors may still require copies.
- No dependency on a global private Inductor configuration in the final SpikingJelly implementation.
